internal/plugins/ansible/templates/watches.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templates
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+
+	"sigs.k8s.io/kubebuilder/pkg/model/file"
+)
+
+var _ file.Template = &Watches{}
+
+const (
+	defaultWatchesFile = "watches.yaml"
+	watchMarker        = "watch"
+)
+
+// Watches scaffolds the watches.yaml file
+type Watches struct {
+	file.TemplateMixin
+}
+
+// SetTemplateDefaults implements input.Template
+func (f *Watches) SetTemplateDefaults() error {
+	if f.Path == "" {
+		f.Path = defaultWatchesFile
+	}
+
+	f.TemplateBody = fmt.Sprintf(watchesTemplate,
+		file.NewMarkerFor(f.Path, watchMarker),
+	)
+	return nil
+}
+
+var _ file.Inserter = &WatchesUpdater{}
+
+type WatchesUpdater struct {
+	file.TemplateMixin
+	file.ResourceMixin
+
+	GeneratePlaybook bool
+	RolesDir         string
+	// ChartPath string
+}
+
+func (*WatchesUpdater) GetPath() string {
+	return defaultWatchesFile
+}
+
+func (*WatchesUpdater) GetIfExistsAction() file.IfExistsAction {
+	return file.Overwrite
+}
+
+func (f *WatchesUpdater) GetMarkers() []file.Marker {
+	return []file.Marker{
+		file.NewMarkerFor(defaultWatchesFile, watchMarker),
+	}
+}
+
+// TODO(asmacdo) Legacy version read the watches file and ensured that duplicate watches were not
+// created. However, creating a new API based on a resource should fail earlier than the watches
+// templating, so for now I am ignoring this case.
+func (f *WatchesUpdater) GetCodeFragments() file.CodeFragmentsMap {
+	fragments := make(file.CodeFragmentsMap, 1)
+
+	// If resource is not being provided we are creating the file, not updating it
+	if f.Resource == nil {
+		return fragments
+	}
+
+	// Generate watch fragments
+	watches := make([]string, 0)
+	buf := &bytes.Buffer{}
+	tmpl := template.Must(template.New("rules").Funcs(file.DefaultFuncMap()).Parse(watchFragment))
+	err := tmpl.Execute(buf, f)
+	if err != nil {
+		panic(err)
```

Comment:
Do you plan on removing this panic? Ideally the watches fragment could be constructed without using a template as to not have to handle an error. If that is not possible/ergonomic, I'd rather see this template executed in the caller and passed as a field to `WatchesUpdater`. Perhaps in an exported function that generates a watches fragment?

Reply:
I had to use a template rather than Sprintf to enable `{{.Resource.Kind | lower }}` in the fragment. I borrowed the error handling from a [similar situation in the helm plugin](https://github.com/operator-framework/operator-sdk/blob/master/internal/plugins/helm/v1/scaffolds/templates/role.go#L106-L110)

I assume the reason for moving this into an exported function is that it would improve error handling?

Reply:
Pretty much. I'd like to avoid panics if possible, and this function technically could panic because it takes user input (Resource), however unlikely due to upstream validation. I'm fine leaving this alone for now with a TODO and addressing in a follow-up.

Reply:
IMO the watches here would be == the helm one. However, instead of you have the Charts attribute you will have the ansible role and playbooks.